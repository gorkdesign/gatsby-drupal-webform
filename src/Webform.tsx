import React, { useState, FormEvent } from 'react'
import axios from 'axios'
import { graphql } from 'gatsby'

import { formToJSON } from './utils'
import { renderWebformElement, DEFAULT_SUBMIT_LABEL } from './webformRender'

/**
 * Webform object as returned from GraphQL query.
 */
export interface WebformObject {
	drupal_internal__id: string
	description: string
	status: string
	elements: WebformElement[]
}

export type WebformOption = {
	label: string
	value: string
}

export interface WebformStateCondition {
	[key: string]: boolean | string | null
}

/**
 * Webform drupal form api states.
 */
export type WebformState = {
	state: string
	selector: string
	condition: WebformStateCondition
}

/**
 * Webform element attribute.
 */
export type WebformAttribute = {
	name: string
	value: string
}

/**
 * Webform element (e.g. text field input or submit button).
 */
export type WebformElement = {
	name: string
	type: string
	options?: WebformOption[]
	attributes: WebformAttribute[]
	states?: WebformState[]
}

export type WebformCustomComponentProps = {
	element: WebformElement
	error?: string
}

/**
 * Custom component for webform element
 */
export type WebformCustomComponent = React.FC<WebformCustomComponentProps>

export type WebformValidateHandler = (event: FormEvent<HTMLFormElement>) => boolean
export type WebformSubmitHandler = (
	data: ReturnType<typeof formToJSON>,
	event: FormEvent<HTMLFormElement>
) => void | boolean | Promise<void | boolean>

export type WebformSuccessHandler = (response: any, event: FormEvent<HTMLFormElement>) => void
export type WebformErrorHandler = (err: any, event: FormEvent<HTMLFormElement>) => void

export class WebformError extends Error {
	response: any

	constructor(response: any) {
		super()

		this.response = response
	}
}

/**
 * Element errors returned by Drupal.
 */
type WebformErrors = {
	[name: string]: string
}

interface Props {
	id?: string
	className?: string
	style?: React.CSSProperties

	webform: WebformObject

	onValidate?: WebformValidateHandler
	onSuccess?: WebformSuccessHandler
	onError?: WebformErrorHandler

	/**
	 * Called right before webform data is POSTed to the api.
	 *
	 * If callback returns false nothing is sent to the api.
	 */
	onSubmit?: WebformSubmitHandler

	/** Form POST endpoint */
	endpoint: string

	/** Provide custom components that handle specific webform elements. */
	customComponents: { [name: string]: WebformCustomComponent }
}

/**
 * Drupal webform react component.
 */
const Webform = ({ webform, customComponents, ...props }: Props) => {
	const [errors, setErrors] = useState<WebformErrors>({})

	const submitHandler: React.FormEventHandler<HTMLFormElement> = async event => {
		event.preventDefault()

		const target = event.currentTarget

		// Clear errors from previous submit.
		setErrors({})

		// Remove lingering css classes from previous submits.
		target.classList.remove('form-submitting', 'form-error', 'form-submitted')

		if ((!props.onValidate || props.onValidate(event)) && target.checkValidity()) {
			// Let css know that this form was validated and is being submitted.
			target.classList.add('was-validated', 'form-submitting')

			// Serialize form data.
			const data = formToJSON(target.elements)

			try {
				// If onSubmit returns false skip submitting to API.
				if (props.onSubmit && (await props.onSubmit(data, event)) === false) {
					target.classList.replace('form-submitting', 'form-submitted')
					return
				}

				// Submit form to API.
				const response = await axios.post(props.endpoint, {
					...data,
					webform_id: webform.drupal_internal__id
				})

				if (response.data.error) {
					throw new WebformError(response)
				}

				// Convey current form state.
				target.classList.replace('form-submitting', 'form-submitted')
				props.onSuccess && props.onSuccess(event, response.data)
			} catch (err) {
				// API should return error structure if validation fails.
				// We use that to render error messages to the form.
				if (err.response && err.response.data.error) {
					setErrors(err.response.data.error)
				}

				// Convey current form state.
				target.classList.replace('form-submitting', 'form-error')
				props.onError && props.onError(err, event)
			}
		} else {
			// Let css know this form was validated.
			target.classList.add('was-validated', 'form-error')
		}
	}

	return (
		<form
			onSubmit={submitHandler}
			id={props.id}
			className={props.className}
			style={props.style}
			data-webform-id={webform.drupal_internal__id}
		>
			{/* Render webform elements */}
			{webform.elements.map(element => (
				<React.Fragment key={element.name}>
					{renderWebformElement(element, errors[element.name], customComponents![element.name])}
				</React.Fragment>
			))}

			{/* Render default submit button if it is not defined in elements array. */}
			{webform.elements.find(element => element.type === 'webform_actions') === undefined && (
				<button type="submit">{DEFAULT_SUBMIT_LABEL}</button>
			)}
		</form>
	)
}

Webform.defaultProps = {
	customComponents: {}
}

export default Webform

export const query = graphql`
	fragment SimpleWebform on webform__webform {
		drupal_internal__id
		elements {
			name
			type
			attributes {
				name
				value
			}
		}
	}
`
