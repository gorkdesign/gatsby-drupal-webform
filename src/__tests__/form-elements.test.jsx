import React from 'react'
import renderer from 'react-test-renderer'

import Webform from '../Webform'

const webform = {
	drupal_internal__id: 'contact',
	elements: [
		{
			name: 'name',
			type: 'textfield',
			attributes: [
				{
					name: '#title',
					value: 'Your Name'
				},
				{
					name: '#required',
					value: 'true'
				},
				{
					name: '#default_value',
					value: 'John'
				}
			]
		},
		{
			name: 'email',
			type: 'email',
			attributes: [
				{
					name: '#title',
					value: 'Your Email'
				},
				{
					name: '#required',
					value: 'true'
				},
				{
					name: '#default_value',
					value: 'john@example.com'
				}
			]
		},
		{
			name: 'subject',
			type: 'email',
			attributes: [
				{
					name: '#title',
					value: 'Subject'
				},
				{
					name: '#required',
					value: 'true'
				}
			]
		},
		{
			name: 'message',
			type: 'textarea',
			attributes: [
				{
					name: '#title',
					value: 'Subject'
				},
				{
					name: '#required',
					value: 'true'
				}
			]
		},
		{
			name: 'actions',
			type: 'webform_actions',
			attributes: [
				{
					name: '#title',
					value: 'Submit button(s)'
				},
				{
					name: '#submit__label',
					value: 'Send message'
				}
			]
		}
	]
}

describe('Webform', () => {
	it('contact form renders correctly', () => {
		const tree = renderer
			.create(<Webform className="my-form" style={{ color: 'black' }} endpoint="/form-submit" webform={webform} />)
			.toJSON()

		expect(tree).toMatchSnapshot()
	})

	it('title attributes', () => {
		const webform = {
			drupal_internal__id: 'contact',
			elements: [
				{
					name: 'name',
					type: 'textfield',
					attributes: [
						{
							name: '#title',
							value: 'Your Name'
						},
						{
							name: '#title_display',
							value: 'invisible'
						}
					]
				}
			]
		}

		const tree = renderer
			.create(<Webform className="my-form" style={{ color: 'black' }} endpoint="/form-submit" webform={webform} />)
			.toJSON()

		expect(tree).toMatchSnapshot()
	})

	it('hidden inputs', () => {
		const webform = {
			drupal_internal__id: 'contact',
			elements: [
				{
					name: 'name',
					type: 'hidden',
					attributes: [
						{
							name: '#title',
							value: 'Your Name'
						}
					]
				}
			]
		}

		const tree = renderer
			.create(<Webform className="my-form" style={{ color: 'black' }} endpoint="/form-submit" webform={webform} />)
			.toJSON()

		expect(tree).toMatchSnapshot()
	})

	it('checkboxes', () => {
		const webform = {
			drupal_internal__id: 'contact',
			elements: [
				{
					name: 'tos',
					type: 'checkbox',
					attributes: [
						{
							name: '#title',
							value: 'Agree terms of service'
						}
					]
				},
				{
					name: 'favourite_color',
					type: 'radios',
					options: [
						{
							label: 'Hotpink',
							value: '#ff69b4'
						},
						{
							label: 'Olive',
							value: '#808000'
						},
						{
							label: 'Cornflower blue',
							value: '#6495ed'
						}
					],
					attributes: [
						{
							name: '#title',
							value: 'Select your favourite color'
						}
					]
				}
			]
		}

		const tree = renderer
			.create(<Webform className="my-form" style={{ color: 'black' }} endpoint="/form-submit" webform={webform} />)
			.toJSON()

		expect(tree).toMatchSnapshot()
	})
})
