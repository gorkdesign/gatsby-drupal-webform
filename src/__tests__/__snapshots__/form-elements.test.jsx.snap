// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webform checkboxes 1`] = `
<form
  className="my-form"
  data-webform-id="contact"
  onSubmit={[Function]}
  style={
    Object {
      "color": "black",
    }
  }
>
  <div
    className="form-group form-check"
  >
    <input
      className="form-check-input"
      defaultChecked={false}
      id="form-tos"
      name="tos"
      type="checkbox"
    />
    <label
      htmlFor="form-tos"
    >
      Agree terms of service
    </label>
  </div>
  <div
    className="form-group"
  >
    <label>
      Select your favourite color
    </label>
    <div
      className="form-check"
    >
      <input
        className="form-check-input"
        defaultChecked={false}
        id="form-favourite-color-#ff69b4"
        name="favourite_color"
        type="radio"
        value="#ff69b4"
      />
      <label
        className="form-check-label"
        htmlFor="form-favourite-color-#ff69b4"
      >
        Hotpink
      </label>
    </div>
    <div
      className="form-check"
    >
      <input
        className="form-check-input"
        defaultChecked={false}
        id="form-favourite-color-#808000"
        name="favourite_color"
        type="radio"
        value="#808000"
      />
      <label
        className="form-check-label"
        htmlFor="form-favourite-color-#808000"
      >
        Olive
      </label>
    </div>
    <div
      className="form-check"
    >
      <input
        className="form-check-input"
        defaultChecked={false}
        id="form-favourite-color-#6495ed"
        name="favourite_color"
        type="radio"
        value="#6495ed"
      />
      <label
        className="form-check-label"
        htmlFor="form-favourite-color-#6495ed"
      >
        Cornflower blue
      </label>
    </div>
  </div>
  <button
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`Webform contact form renders correctly 1`] = `
<form
  className="my-form"
  data-webform-id="contact"
  onSubmit={[Function]}
  style={
    Object {
      "color": "black",
    }
  }
>
  <div
    className="form-group"
  >
    <label
      htmlFor="form-name"
    >
      Your Name
    </label>
    <input
      className="form-control"
      defaultValue="John"
      id="form-name"
      name="name"
      required="true"
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="form-email"
    >
      Your Email
    </label>
    <input
      className="form-control"
      defaultValue="john@example.com"
      id="form-email"
      name="email"
      required="true"
      type="email"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="form-subject"
    >
      Subject
    </label>
    <input
      className="form-control"
      id="form-subject"
      name="subject"
      required="true"
      type="email"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="form-message"
    >
      Subject
    </label>
    <textarea
      className="form-control"
      id="form-message"
      name="message"
      required="true"
    />
  </div>
  <div
    className="form-group"
  >
    <button
      type="submit"
    >
      Send message
    </button>
  </div>
</form>
`;

exports[`Webform hidden inputs 1`] = `
<form
  className="my-form"
  data-webform-id="contact"
  onSubmit={[Function]}
  style={
    Object {
      "color": "black",
    }
  }
>
  <input
    className="form-control"
    id="form-name"
    name="name"
    type="hidden"
  />
  <button
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`Webform title attributes 1`] = `
<form
  className="my-form"
  data-webform-id="contact"
  onSubmit={[Function]}
  style={
    Object {
      "color": "black",
    }
  }
>
  <div
    className="form-group"
  >
    <label
      htmlFor="form-name"
      style={
        Object {
          "clip": "rect(1px,1px,1px,1px)",
          "height": "1px",
          "overflow": "hidden",
          "position": "absolute",
          "width": "1px",
          "wordWrap": "normal",
        }
      }
    >
      Your Name
    </label>
    <input
      className="form-control"
      id="form-name"
      name="name"
      type="text"
    />
  </div>
  <button
    type="submit"
  >
    Submit
  </button>
</form>
`;
